File: apps/ios/OpenFrame/Repository/TaskRepository.swift
import Foundation

final class TaskRepository: Sendable {
    private let apiClient: APIClient

    init(apiClient: APIClient) {
        self.apiClient = apiClient
    }

    func getTaskLists() async -> Result<[TaskList], Error> {
        do {
            let dtos: [TaskListDTO] = try await apiClient.requestData(.get, path: "/api/v1/tasks/lists")
            return .success(dtos.map { $0.toDomain() })
        } catch {
            return .failure(error)
        }
    }

    func getTasks(listId: String? = nil, status: String? = nil) async -> Result<[OFTask], Error> {
        do {
            var query: [String: String] = [:]
            if let listId { query["listId"] = listId }
            if let status { query["status"] = status }
            let dtos: [TaskDTO] = try await apiClient.requestData(.get, path: "/api/v1/tasks", query: query.isEmpty ? nil : query)
            return .success(dtos.map { $0.toDomain() })
        } catch {
            return .failure(error)
        }
    }

    func createTask(taskListId: String, title: String, notes: String? = nil, dueDate: String? = nil) async -> Result<OFTask, Error> {
        do {
            let body = CreateTaskRequest(taskListId: taskListId, title: title, notes: notes, dueDate: dueDate)
            let dto: TaskDTO = try await apiClient.requestData(.post, path: "/api/v1/tasks", body: body)
            return .success(dto.toDomain())
        } catch {
            return .failure(error)
        }
    }

    func completeTask(id: String) async -> Result<OFTask, Error> {
        do {
            let dto: TaskDTO = try await apiClient.requestData(.post, path: "/api/v1/tasks/\(id)/complete")
            return .success(dto.toDomain())
        } catch {
            return .failure(error)
        }
    }

    func deleteTask(id: String) async -> Result<Void, Error> {
        do {
            try await apiClient.requestVoid(.delete, path: "/api/v1/tasks/\(id)")
            return .success(())
        } catch {
            return .failure(error)
        }
    }
}

File: apps/android/app/src/main/java/us/openframe/app/data/repository/TaskRepository.kt
package us.openframe.app.data.repository

import us.openframe.app.data.remote.api.TaskApi
import us.openframe.app.data.remote.dto.CreateTaskRequest
import us.openframe.app.data.remote.dto.TaskDto
import us.openframe.app.data.remote.dto.TaskListDto
import us.openframe.app.domain.model.Task
import us.openframe.app.domain.model.TaskList
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class TaskRepository @Inject constructor(
    private val taskApi: TaskApi,
) {
    suspend fun getTaskLists(): Result<List<TaskList>> {
        return try {
            val response = taskApi.getTaskLists()
            if (response.isSuccessful) {
                val lists = (response.body()?.data ?: emptyList()).map { it.toDomain() }
                Result.success(lists)
            } else {
                Result.failure(Exception("Failed to fetch task lists"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun getTasks(listId: String? = null, status: String? = null): Result<List<Task>> {
        return try {
            val response = taskApi.getTasks(listId, status)
            if (response.isSuccessful) {
                val tasks = (response.body()?.data ?: emptyList()).map { it.toDomain() }
                Result.success(tasks)
            } else {
                Result.failure(Exception("Failed to fetch tasks"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun createTask(taskListId: String, title: String, notes: String? = null): Result<Task> {
        return try {
            val response = taskApi.createTask(CreateTaskRequest(taskListId, title, notes))
            if (response.isSuccessful) {
                val task = response.body()?.data?.toDomain() ?: return Result.failure(Exception("No data"))
                Result.success(task)
            } else {
                Result.failure(Exception("Failed to create task"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun completeTask(id: String): Result<Task> {
        return try {
            val response = taskApi.completeTask(id)
            if (response.isSuccessful) {
                val task = response.body()?.data?.toDomain() ?: return Result.failure(Exception("No data"))
                Result.success(task)
            } else {
                Result.failure(Exception("Failed to complete task"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun deleteTask(id: String): Result<Unit> {
        return try {
            val response = taskApi.deleteTask(id)
            if (response.isSuccessful) Result.success(Unit)
            else Result.failure(Exception("Failed to delete task"))
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}

private fun TaskListDto.toDomain() = TaskList(id = id, name = name, isVisible = isVisible)

private fun TaskDto.toDomain() = Task(
    id = id, taskListId = taskListId, title = title, notes = notes,
    status = status, dueDate = dueDate, completedAt = completedAt, listName = list?.name,
)
